<a href="https://colab.research.google.com/github/JD12321/2023-DA/blob/main/005_KNN_P069(%ED%99%9C%EC%9A%A9).ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# KNN(K-Nearest Neighbors)


## import library & read csv data

In [ ]:
# library & data
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from sklearn import metrics
# from sklearn.datasets import load_iris
# iris = load_iris()

# column of dataset
#names = ['Sepal.Length', 'Sepal.Width', 'Petal.Length', 'Petal.Width', 'Species']
#dataset = pd.read_csv('/content/dgu_data/Iris.csv')

#== [pd.read_csv] 부분이 빈칸으로 되어 있고 쓰는 시험 문제 나옴 ==#
dataset = pd.read_csv('/content/sample_data/Data_all/Iris.csv')

## search of data

In [ ]:
#== 아래 10개 코드는 시험에 많이 나옴 (암기) - 2문제 출제 ==#

#dataset
#print(dataset)
#dataset.info()
#dataset.describe()
#dataset.head(5)        # 안써도 5개 나옴 - a.head()
#dataset.tail()            # 안써도 5개 나옴 - a.tail()
#dataset.columns       # dataset.rows는 없음
#dataset.shape
#dataset.dtypes
#dataset.isnull().sum()  # null 값이 없는게 좋은 데이터임

In [ ]:
# training과 test dataset 분리 (슬라이싱 값 수정)
#= 슬라이싱 하는거 시험 나옴 =#
X = dataset.iloc[:, 1:-1].values    # 뒤에서 하나를 뺀 값을 가져와서 X에 저장
y = dataset.iloc[:, 5].values        # 열은 앞에서 다섯 번째 값만 가져와서 y에 저장

X.shape
#X
#y.shape
#y
#print(X)
#print(y)
#print(X, y)
print(X.shape, y.shape)

(150, 4) (150,)


## training dataset : 80% / test dataset : 20%

In [ ]:
#== 이 코드 블럭 전체를 빈칸으로 두고 다 쓰는게 시험에 나옴 (다 암기) ==#

from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.15)   ### 20% --> 15% ###

#X_train.shape
#X_test.shape
#y_train.shape
y_test.shape


(45,)

In [ ]:
# Data Standardization(표준화) : Feature의 범위가 다른 경우에만 사용
from sklearn.preprocessing import StandardScaler
s = StandardScaler()   # 특성 스케일링(scaling), 평균이 0, 표준편차가 1이 되도록 변환
s.fit(X_train, X_test)              ## 교과서를 변경한 부분

X_train = s.transform(X_train)    # 훈련 데이터를 스케일링 처리
X_test = s.transform(X_test)      # 테스트 데이터를 스케일링 처리

#X_train
X_test

array([[ 0.11758258, -0.0685618 ,  0.51606242,  0.7155277 ],
       [ 0.49942292, -1.26839335,  0.57633668,  0.30646921],
       [-1.02793845,  0.65133712, -1.35243946, -1.05705907],
       [-0.26425776,  3.29096653, -1.47298797, -1.19341189],
       [-0.90065833,  1.13126974, -1.47298797, -1.46611755],
       [-0.6460981 ,  2.09113498, -1.35243946, -1.19341189],
       [-0.00969753, -0.0685618 ,  0.69688519,  0.7155277 ],
       [ 0.88126327, -0.0685618 ,  0.27496541,  0.17011639],
       [ 2.15406441, -0.0685618 ,  1.600999  ,  1.12458618],
       [-0.00969753, -0.0685618 ,  0.1544169 ,  0.30646921],
       [-1.15521856,  0.41137082, -1.65381073, -1.46611755],
       [ 2.28134452, -0.0685618 ,  1.29962773,  1.39729184],
       [-1.66433902,  1.37123605, -1.77435924, -1.46611755],
       [-1.28249867, -1.50835966, -0.38805139, -0.37529493],
       [-0.51881799, -1.50835966, -0.08668012, -0.2389421 ],
       [-1.66433902,  0.89130343, -1.53326222, -1.32976472],
       [ 0.2448627 , -0.

In [ ]:
# create model (KNN)
from sklearn.neighbors import KNeighborsClassifier

### k = 50  -->  k = 13 ###
knn = KNeighborsClassifier(n_neighbors=13)      # K의 값 : 3,5,7,9,11,13,15,17,19...,49
#y_train = y_train.astype(int)    # fix type error
knn.fit(X_train, y_train)

KNeighborsClassifier(n_neighbors=19)

In [ ]:
# model accuracy
from sklearn import metrics
y_pred = knn.predict(X_test)
#y_tset = y_test.astype(int)
print("Accuracy: {}".format(metrics.accuracy_score(y_test, y_pred)))

Accuracy: 0.9333333333333333


In [ ]:
#########################################
k = 10
acc_array = np.zeros(k)
for k in np.arange(1, k+1, 1):
    classifier = KNeighborsClassifier(n_neighbors=k).fit(X_train, y_train)
    y_pred = knn.predict(X_test)
    acc = metrics.accuracy_score(y_test, y_pred)
    acc_array[k-1] = acc

max_acc = np.amax(acc_array)
acc_list = list(acc_array)
k = acc_list.index(max_acc)
print("Accuracy is", metrics.accuracy_score(y_test, y_pred), "for K-Value:", k)

Accuracy is 0.9333333333333333 for K-Value: 0


---
# **최적의 하이퍼파라미터 테스트 코드**


In [ ]:
# library & data
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from sklearn import metrics
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.neighbors import KNeighborsClassifier
from sklearn import metrics

x2 = []
dataset = pd.read_csv('/content/sample_data/Data_all/Iris.csv')

X = dataset.iloc[:, 1:-1].values
y = dataset.iloc[:, 5].values

for i in range(10000):

  X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.05)
  y_test.shape

  s = StandardScaler()
  s.fit(X_train, X_test)

  X_train = s.transform(X_train)
  X_test = s.transform(X_test)

  knn = KNeighborsClassifier(n_neighbors=13)
  knn.fit(X_train, y_train)

  y_pred = knn.predict(X_test)
  x1 = metrics.accuracy_score(y_test, y_pred)
  x2.append(x1)

x3 = sum(x2) / len(x2)
print("Accuracy: {}".format(x3))


Accuracy: 0.96365


# **1. testdataset 비율 테스트 결과** (최적 비율 : 15%)

- **고민** : 기준 없이 테스트 데이터를 계속 줄이자니 과적합이 걱정되어, 얼마나 줄여야할지를 모르겠음. (그래프를 안그리니 elbow를 찾기 힘듬)

- **5%** : 0.96365

- **10%** : 0.9629399999999778

- **15%** : 0.9619130434781661

- **20%** : 0.9456966666665878

- **25%** : 0.9440263157895159

- **30%** : 0.9406533333333584

- **35%** : 0.9338867924528371

- **40%** : 0.9255416666666307

# **2. n_neighbors값 테스트 결과** (최적값 : 13)

- **3** : 0.9457533333333382

- **4** : 0.9435333333333387

- **5** : 0.950393333333334

- **6** : 0.9509400000000016

- **7** : 0.9539733333333319

- **8** : 0.9538399999999979

- **9** : 0.9526866666666663

- **10** : 0.9539133333333315

- **11** : 0.9579399999999967

- **12** : 0.9579333333333311

- **13** : 0.959899999999994

- **14** : 0.9571399999999967

- **15** : 0.9554399999999993

- **16** : 0.9522133333333328

- **17** : 0.9494333333333348

- **18** : 0.9471866666666708

- **19** : 0.9471133333333364

- **20** : 0.9461666666666714

